### ISSUE TITLE
Wheel version numbers are incompatible with `torch (==1.11.0)`

### ISSUE THREAD

**Issue by nalzok** [pytorch/xla#3661]
## 🐛 Bug

To illustrate, the version number reported by the following wheels is `1.11.0a0+gitbc2c6ed`.

+ https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl
+ https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-1.11-cp38-cp38-linux_x86_64.whl

This causes problems when I want to install `torchvision (==0.12.0)` because `1.11.0a0+gitbc2c6ed` is incompatible with `torch (==1.11.0)`. We can make it compatible by using `1.11.0.a0+gitbc2c6ed` instead (note the additional dot).

## To Reproduce

```bash
$ pip install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-1.11-cp38-cp38-linux_x86_64.whl torchvision==0.12.0
Collecting torch==1.11
  Downloading https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl (104.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.0/104.0 MB 9.0 MB/s eta 0:00:00
Collecting torch-xla==1.11
  Downloading https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-1.11-cp38-cp38-linux_x86_64.whl (156.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.7/156.7 MB 6.3 MB/s eta 0:00:00
Collecting torchvision==0.12.0
  Downloading torchvision-0.12.0-cp38-cp38-manylinux1_x86_64.whl (21.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.0/21.0 MB 63.5 MB/s eta 0:00:00
Collecting typing-extensions
  Downloading typing_extensions-4.2.0-py3-none-any.whl (24 kB)
Collecting pillow!=8.3.*,>=5.3.0
  Downloading Pillow-9.1.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 72.8 MB/s eta 0:00:00
Collecting numpy
  Downloading numpy-1.22.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.9/16.9 MB 73.3 MB/s eta 0:00:00
Collecting requests
  Downloading requests-2.28.0-py3-none-any.whl (62 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.8/62.8 kB 8.5 MB/s eta 0:00:00
ERROR: Cannot install torch 1.11.0a0+gitbc2c6ed (from https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl) and torchvision==0.12.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested torch 1.11.0a0+gitbc2c6ed (from https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl)
    torchvision 0.12.0 depends on torch==1.11.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

## Expected behavior

Make the version number `1.11.0.a0+gitbc2c6ed`, so that it is compatible with `== 1.11.0` since `.a0+gitbc2c6ed` is a suffix. See [this SO answer](https://stackoverflow.com/questions/64319173/pre-release-versions-are-not-matched-by-pip-when-using-the-pre-option) for more information.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: N/A
 - torch_xla version: `1.11.0.a0+gitbc2c6ed`

**Comment by nalzok** [comment]
I wrote a bash script to do the monkey patching. It essentially replace all occurrence of `torch-1.11.0a0+gitbc2c6ed-cp38-cp38-linux_x86_64.whl` to `torch-1.11.0+a0gitbc2c6ed-cp38-cp38-linux_x86_64.whl`. Of course this approach is terribly hacky, but I assume it's still easier than recompiling the wheel from scratch?

```bash
#!/usr/bin/env bash
set -xeuo pipefail

PACKAGE_DIR=/mnt/disks/persist/cache/storage.googleapis.com/tpu-pytorch/wheels/tpuvm/
PACKAGE_NAME=torch-1.11-cp38-cp38-linux_x86_64.whl

mkdir -p /tmp/surgery
WORKDIR=$(mktemp -d -p /tmp/surgery)

echo Unpacking $PACKAGE_DIR/$PACKAGE_NAME...
unzip -q $PACKAGE_DIR/$PACKAGE_NAME -d $WORKDIR

pushd $WORKDIR
VERSION_OLD=$(ag --nofilename '^Version: ' | cut -d' ' -f2)
VERSION_OLD_ESCAPED=$(echo $VERSION_OLD | sed -e 's/+/\\+/g' -e 's/\./\\./g')
VERSION_NEW=$(echo $VERSION_OLD | sed -e 's/+//g' -e 's/\([a-z].*\)/+\1/g')
find . -name "*$VERSION_OLD_ESCAPED*" | xargs -I {} rename "s/$VERSION_OLD_ESCAPED/$VERSION_NEW/g" {} \;
ag -l -Q $VERSION_OLD | xargs -I {} sed -i "s/$VERSION_OLD/$VERSION_NEW/g" {}
ag -l -Q $VERSION_NEW

popd
pipenv run wheel pack $WORKDIR --dest-dir $PACKAGE_DIR

rm -rf $WORKDIR
```

**Comment by JackCaoG** [comment]
directly using torch's 1.11 wheel works for me
```
jackcao@t1v-n-b7d25fee-w-0:~$ sudo pip3 install torch==1.11.0
Collecting torch==1.11.0
  Downloading torch-1.11.0-cp38-cp38-manylinux1_x86_64.whl (750.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 750.6/750.6 MB 1.2 MB/s eta 0:00:00
Requirement already satisfied: typing-extensions in /usr/local/lib/python3.8/dist-packages (from torch==1.11.0) (4.2.0)
Installing collected packages: torch
Successfully installed torch-1.11.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
jackcao@t1v-n-b7d25fee-w-0:~$ sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-1.11-cp38-cp38-linux_x86_64.whl
Collecting torch-xla==1.11
  Downloading https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-1.11-cp38-cp38-linux_x86_64.whl (156.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.7/156.7 MB 6.2 MB/s eta 0:00:00
Installing collected packages: torch-xla
Successfully installed torch-xla-1.11
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
jackcao@t1v-n-b7d25fee-w-0:~$ sudo pip3 install torchvision==0.12.0
Collecting torchvision==0.12.0
  Downloading torchvision-0.12.0-cp38-cp38-manylinux1_x86_64.whl (21.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.0/21.0 MB 13.3 MB/s eta 0:00:00
Requirement already satisfied: torch==1.11.0 in /usr/local/lib/python3.8/dist-packages (from torchvision==0.12.0) (1.11.0)
Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /usr/local/lib/python3.8/dist-packages (from torchvision==0.12.0) (9.1.1)
Requirement already satisfied: typing-extensions in /usr/local/lib/python3.8/dist-packages (from torchvision==0.12.0) (4.2.0)
Requirement already satisfied: requests in /usr/local/lib/python3.8/dist-packages (from torchvision==0.12.0) (2.28.0)
Requirement already satisfied: numpy in /usr/local/lib/python3.8/dist-packages (from torchvision==0.12.0) (1.22.4)
Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.8/dist-packages (from requests->torchvision==0.12.0) (3.3)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.8/dist-packages (from requests->torchvision==0.12.0) (1.26.9)
Requirement already satisfied: charset-normalizer~=2.0.0 in /usr/local/lib/python3.8/dist-packages (from requests->torchvision==0.12.0) (2.0.12)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.8/dist-packages (from requests->torchvision==0.12.0) (2022.5.18.1)
Installing collected packages: torchvision
  Attempting uninstall: torchvision
    Found existing installation: torchvision 0.14.0a0+c02d6ce
    Uninstalling torchvision-0.14.0a0+c02d6ce:
      Successfully uninstalled torchvision-0.14.0a0+c02d6ce
Successfully installed torchvision-0.12.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
```

btw 1.11 has some speed regression on `all_reduce`. I would recommend 1.10 or we will release 1.12 in next week(whenever pytorch decide to publish their 1.12)

**Comment by nalzok** [comment]
Interesting! What's the difference between the official torch-1.11 and https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl, and why do you release the latter when the former seems to do the job?

Additionally, could you make sure you don't accidentally build against a pre-release version when releasing 1.12? I guess this is the case for `1.11.0a0+gitbc2c6ed`, since `1.11.0a0` seems to be an alpha version.

**Comment by JackCaoG** [comment]
https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl is just a test wheel we used before pytorch offically released their 1.11 wheel. It is not meant to be used by the user.

**Comment by nalzok** [comment]
Thanks for the clarification. I just realized that https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-1.11-cp38-cp38-linux_x86_64.whl wasn't even listed in README. Not sure how I was made aware of its existence and somehow assumed that I am supposed to install that as opposed to `pip3 install torch==1.11.0`.

I guess this is not a bug then. Closing!

**Comment by bones0** [comment]
Ran into the same problem with what I compiled myself. See https://discuss.pytorch.org/t/self-compiled-versions-get-numbered-the-way-they-dont-fit-requirements-of-standard-wheels/186920/2. There is no official and reasonably new torch wheel supporting my Tesla K40c so I have to do it myself. Is the script above still the only possibility to fix it? Or do I have to look into the inner workings of python setup.py bdist_wheel? 

Options for 'bdist_wheel' command:
  --bdist-dir (-b)  temporary directory for creating the distribution
  --plat-name (-p)  platform name to embed in generated filenames (default:
                    win_amd64)
  --keep-temp (-k)  keep the pseudo-installation tree around after creating
                    the distribution archive
  --dist-dir (-d)   directory to put final built distributions in
  --skip-build      skip rebuilding everything (for testing/debugging)
  --relative        build the archive using relative paths (default: false)
  --owner (-u)      Owner name used when creating a tar file [default: current
                    user]
  --group (-g)      Group name used when creating a tar file [default: current
                    group]
  --universal       make a universal wheel (default: false)
  --compression     zipfile compression (one of: stored, deflated) (default:
                    'deflated')
  --python-tag      Python implementation compatibility tag (default: 'py3')
  --build-number    Build number for this particular version. As specified in
                    PEP-0427, this must start with a digit. [default: None]
  --py-limited-api  Python tag (cp32|cp33|cpNN) for abi3 wheel tag (default:
                    false)
                    
Not sure which option might do what I need. --universal is the only option where I am not sure what it's doing so it might be what I need. or not.
